extensions/modules/file/src/main/java/org/exist/xquery/modules/file/Sync.java

```diff
@@ -55,183 +62,356 @@
 import org.exist.xmldb.XmldbURI;
 import org.exist.xquery.BasicFunction;
 import org.exist.xquery.Cardinality;
+import org.exist.xquery.ErrorCodes;
 import org.exist.xquery.FunctionSignature;
 import org.exist.xquery.XPathException;
 import org.exist.xquery.XQueryContext;
-import org.exist.xquery.value.DateTimeValue;
-import org.exist.xquery.value.FunctionParameterSequenceType;
-import org.exist.xquery.value.FunctionReturnSequenceType;
-import org.exist.xquery.value.Sequence;
-import org.exist.xquery.value.SequenceType;
-import org.exist.xquery.value.Type;
+import org.exist.xquery.functions.map.AbstractMapType;
+import org.exist.xquery.value.*;
 import org.exist.xslt.TransformerFactoryAllocator;
 import org.w3c.dom.Document;
 import org.xml.sax.SAXException;
+import uk.ac.ic.doc.slurp.multilock.MultiLock;
 
 public class Sync extends BasicFunction {
 
-	public final static FunctionSignature signature =
-		new FunctionSignature(
-            new QName("sync", FileModule.NAMESPACE_URI, FileModule.PREFIX),
-            "Synchronize a collection with a directory hierarchy. Compares last modified time stamps. " +
-            "If $dateTime is given, only resources modified after this time stamp are taken into account. " +
-    		"This method is only available to the DBA role.",
-            new SequenceType[]{
-            	new FunctionParameterSequenceType("collection", Type.STRING, 
-                        Cardinality.EXACTLY_ONE, "The collection to sync."),
-                new FunctionParameterSequenceType("targetPath", Type.ITEM, 
-                        Cardinality.EXACTLY_ONE, "The full path or URI to the directory"),
-                new FunctionParameterSequenceType("dateTime", Type.DATE_TIME, 
-                        Cardinality.ZERO_OR_ONE,
-                		"Optional: only resources modified after the given date/time will be synchronized.")
-            },
-            new FunctionReturnSequenceType(Type.BOOLEAN, Cardinality.EXACTLY_ONE, "true if successful, false otherwise")
-        );
-	
-	private final static Properties DEFAULT_PROPERTIES = new Properties();
-	static {
-		DEFAULT_PROPERTIES.put(OutputKeys.INDENT, "yes");
-		DEFAULT_PROPERTIES.put(OutputKeys.OMIT_XML_DECLARATION, "no");
+    public static final String PRUNE_OPT = "prune";
+    public static final String AFTER_OPT = "after";
+    public static final String EXCLUDES_OPT = "excludes";
+
+    public static final QName FileSyncElement = new QName("sync", FileModule.NAMESPACE_URI);
+    public static final QName FileUpdateElement = new QName("update", FileModule.NAMESPACE_URI);
+    public static final QName FileDeleteElement = new QName("delete", FileModule.NAMESPACE_URI);
+    public static final QName FileErrorElement = new QName("error", FileModule.NAMESPACE_URI);
+
+    // TODO(JL) Figure out which namespace all attributes should be in (possible breaking change)
```

Comment:
Is this TODO still valid or have you figured it out?
If this is something to improve in future, perhaps it would be good to open an issue and add the link to the TODO.